metals/src/main/scala/scala/meta/internal/pantsbuild/BloopPants.scala

```diff
@@ -0,0 +1,657 @@
+package scala.meta.internal.pantsbuild
+
+import bloop.config.{Config => C}
+import java.nio.file.Paths
+import java.nio.file.Files
+import java.nio.charset.StandardCharsets
+import java.nio.file.Path
+import scala.collection.mutable
+import ujson.Value
+import scala.util.Success
+import scala.util.Failure
+import scala.util.Try
+import scala.meta.internal.metals.BuildInfo
+import scala.meta.internal.metals.Timer
+import scala.meta.internal.metals.Time
+import java.nio.file.NoSuchFileException
+import scala.meta.internal.mtags.MD5
+import scala.util.Properties
+import coursierapi.Dependency
+import scala.concurrent.ExecutionContext
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.process.SystemProcess
+import scala.meta.pc.CancelToken
+import scala.util.control.NonFatal
+import scala.meta.internal.pc.InterruptException
+import scala.meta.internal.metals.MetalsLogger
+import scala.meta.io.AbsolutePath
+import scala.meta.internal.ansi.LineListener
+import java.util.concurrent.CancellationException
+import ch.epfl.scala.bsp4j.BuildTargetIdentifier
+import scala.sys.process.Process
+
+object BloopPants {
+
+  def main(argStrings: Array[String]): Unit = {
+    MetalsLogger.updateDefaultFormat()
+    Args.parse(argStrings.toList) match {
+      case Left(errors) =>
+        errors.foreach { error =>
+          scribe.error(error)
+        }
+        System.exit(1)
+      case Right(args) =>
+        if (args.isHelp) {
+          println(args.helpMessage)
+        } else if (args.isRegenerate) {
+          bloopRegenerate(
+            AbsolutePath(args.workspace),
+            args.targets
+          )(ExecutionContext.global)
+        } else {
+          val workspace = args.workspace
+          val targets = args.targets
+          val timer = new Timer(Time.system)
+          val installResult = bloopInstall(args)(ExecutionContext.global)
+          installResult match {
+            case Failure(exception) =>
+              scribe.error(s"bloopInstall failed in $timer", exception)
+              sys.exit(1)
+            case Success(count) =>
+              scribe.info(s"time: exported ${count} Pants target(s) in $timer")
+          }
+        }
+    }
+  }
+
+  def bloopAddOwnerOf(
+      workspace: AbsolutePath,
+      source: AbsolutePath
+  ): Seq[BuildTargetIdentifier] = synchronized {
+    val targets = pantsOwnerOf(workspace, source)
+    if (targets.nonEmpty) {
+      val bloopDir = workspace.resolve(".bloop")
+      for {
+        target <- targets
+        jsonFile = bloopDir.resolve(BloopPants.makeJsonFilename(target))
+        if jsonFile.isFile
+      } {
+        val json = ujson.read(jsonFile.readText)
+        val sources = json("project")("sources").arr
+        val sourceStr = Value.Str(source.toString())
+        if (!sources.contains(sourceStr)) {
+          sources += sourceStr
+          jsonFile.writeText(ujson.write(json, indent = 4))
```

Comment:
I believe Jorge was strongly opposed to anything than bloop modifying the config file while it's working :thinking:

Reply:
The config file can be modified while it's working, but it must be updated atomically. I don't remember being strongly opposed to this, but I do remember mentioning that tools that don't own these configuration files should never update them, that's something that only the original build tool can do. Let me know if this clarifies your thinking @tgodzik :smile:

Reply:
@jvican Thanks! That makes sense, I must have misunderstood previously.

Reply:
Sounds great! I will follow up with a PR to make our file write operations atomic.

We use `_root_.bloop.config.write(json, out)` to write most of the JSON files in the Pants export. This step here happens only when the user creates a new file and we need to re-expand the file globs.